expression/builtin_miscellaneous.go

```diff
@@ -176,23 +183,130 @@ func (c *anyValueFunctionClass) getFunction(args []Expression, ctx context.Conte
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinAnyValueSig{newBaseBuiltinFunc(args, ctx)}
+	argTp := fieldTp2EvalTp(args[0].GetType())
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, argTp, argTp)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	*bf.tp = *args[0].GetType()
+	var sig builtinFunc
+	switch argTp {
+	case tpDecimal:
+		sig = &builtinDecimalAnyValueSig{baseDecimalBuiltinFunc{bf}}
+	case tpDuration:
+		sig = &builtinDurationAnyValueSig{baseDurationBuiltinFunc{bf}}
+	case tpInt:
+		sig = &builtinIntAnyValueSig{baseIntBuiltinFunc{bf}}
+	case tpJSON:
+		sig = &builtinJSONAnyValueSig{baseJSONBuiltinFunc{bf}}
+	case tpReal:
+		sig = &builtinRealAnyValueSig{baseRealBuiltinFunc{bf}}
+	case tpString:
+		sig = &builtinStringAnyValueSig{baseStringBuiltinFunc{bf}}
+	case tpTime:
+		sig = &builtinTimeAnyValueSig{baseTimeBuiltinFunc{bf}}
+	default:
+		panic("unexpected evalTp of builtin function ANY_VALUE")
+	}
 	return sig.setSelf(sig), nil
 }
 
-type builtinAnyValueSig struct {
-	baseBuiltinFunc
+type builtinDecimalAnyValueSig struct {
+	baseDecimalBuiltinFunc
 }
 
-// eval evals a builtinAnyValueSig.
+// evalDecimal evals a builtinDecimalAnyValueSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_any-value
-func (b *builtinAnyValueSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
+func (b *builtinDecimalAnyValueSig) evalDecimal(row []types.Datum) (*types.MyDecimal, bool, error) {
+	val, isNull, err := b.args[0].EvalDecimal(row, b.ctx.GetSessionVars().StmtCtx)
+	if err != nil || isNull {
```

Comment:
I mean that,
we do not need to check `if err != nil || isNull`,
we can `return val, isNull, errors.Trace(err)` directly.